src/internal/connector/graph/service.go

```diff
@@ -294,3 +296,26 @@ func (handler *LoggingMiddleware) Intercept(
 
 	return resp, err
 }
+
+// Run a function with retries
+func RunWithRetry(run func() error) error {
+	var err error
+
+	for i := 0; i < numberOfRetries; i++ {
+		err = run()
+		if err == nil {
+			return nil
+		}
+
+		// only retry on timeouts and 500-internal-errors.
+		if !(IsErrTimeout(err) || IsInternalServerError(err)) {
+			break
+		}
+
+		if i < numberOfRetries {
+			time.Sleep(time.Duration(3*(i+2)) * time.Second)
+		}
+	}
+
+	return support.ConnectorStackErrorTraceWrap(err, "maximum retries or unretryable")
```

Comment:
The test failure highlights an issue here. If we change the error returned and do not actually return the *last* Graph API error - error handling in callers can break.

From [test](https://github.com/alcionai/corso/actions/runs/4087123550/jobs/7047752182):
```
  2023-02-03T19:53:13.139Z	INFO	deserialzing previous backup metadata	{"num_collections": 0}
  2023-02-03T19:53:13.397Z	ERROR	non-2xx graph api response	{"method": "GET", "status": "400 Bad Request", "statusCode": 400, "requestLen": 0, "url": "https://graph.microsoft.com/v1.0/users/testevents%408qzvrj.onmicrosoft.com/drives", "response": "HTTP/1.1 400 Bad Request\r\nTransfer-Encoding: chunked\r\nClient-Request-Id: 8c815bdd-abfe-4f58-acc2-2df88b4af604\r\nContent-Type: application/json\r\nDate: Fri, 03 Feb 2023 19:53:12 GMT\r\nRequest-Id: c1ac5431-ff93-4ade-98c2-cfe75218443c\r\nStrict-Transport-Security: max-age=31536000\r\nVary: Accept-Encoding\r\nX-Ms-Ags-Diagnostic: {\"ServerInfo\":{\"DataCenter\":\"East US\",\"Slice\":\"E\",\"Ring\":\"5\",\"ScaleUnit\":\"005\",\"RoleInstance\":\"MN1PEPF0000958D\"}}\r\n\r\nec\r\n{\"error\":{\"code\":\"BadRequest\",\"message\":\"Unable to retrieve user's mysite URL.\",\"innerError\":{\"date\":\"2023-02-03T19:53:13\",\"request-id\":\"c1ac5431-ff93-4ade-98c2-cfe75218443c\",\"client-request-id\":\"8c815bdd-abfe-4f58-acc2-2df88b4af604\"}}}\r\n0\r\n\r\n"}
      drive_test.go:467: 
          	Error Trace:	/home/runner/work/corso/corso/src/internal/connector/onedrive/drive_test.go:467
          	Error:      	Received unexpected error:
          	            	error status code received from the API
          	            	maximum retries or unretryable:  BadRequest Unable to retrieve user's mysite URL.
          	            	github.com/alcionai/corso/src/internal/connector/support.ConnectorStackErrorTraceWrap
          	            		/home/runner/work/corso/corso/src/internal/connector/support/errors.go:98
          	            	github.com/alcionai/corso/src/internal/connector/graph.RunWithRetry
          	            		/home/runner/work/corso/corso/src/internal/connector/graph/service.go:320
          	            	github.com/alcionai/corso/src/internal/connector/onedrive/api.(*userDrivePager).GetPage
          	            		/home/runner/work/corso/corso/src/internal/connector/onedrive/api/drive.go:117
          	            	github.com/alcionai/corso/src/internal/connector/onedrive.drives
          	            		/home/runner/work/corso/corso/src/internal/connector/onedrive/drive.go:80
          	            	github.com/alcionai/corso/src/internal/connector/onedrive.(*Collections).Get
          	            		/home/runner/work/corso/corso/src/internal/connector/onedrive/collections.go:267
          	            	github.com/alcionai/corso/src/internal/connector/onedrive.(*OneDriveSuite).TestOneDriveNewCollections.func1
```